server/src/main/java/org/opensearch/index/remote/RemoteIndexPathUploader.java

```diff
@@ -0,0 +1,267 @@
+                indexUUID,
+                getUploadPathLatchedActionListener(idxMD, latch, exceptionList, SEGMENT_PATH)
+            );
+        }
+    }
+
+    private Repository validateAndGetRepository(String repoSetting) {
+        final String repo = settings.get(repoSetting);
+        assert repo != null : "Remote " + repoSetting + " repository is not configured";
+        final Repository repository = repositoriesService.get().repository(repo);
+        assert repository instanceof BlobStoreRepository : "Repository should be instance of BlobStoreRepository";
+        return repository;
+    }
+
+    public void start() {
+        assert isRemoteStoreClusterStateEnabled(settings) == true : "Remote cluster state is not enabled";
+        if (isRemoteDataAttributePresent == false) {
+            // If remote store data attributes are not present than we skip this.
+            return;
+        }
+        translogRepository = (BlobStoreRepository) validateAndGetRepository(TRANSLOG_REPO_NAME_KEY);
+        segmentRepository = (BlobStoreRepository) validateAndGetRepository(SEGMENT_REPO_NAME_KEY);
+    }
+
+    private boolean isTranslogSegmentRepoSame() {
+        String translogRepoName = settings.get(TRANSLOG_REPO_NAME_KEY);
+        String segmentRepoName = settings.get(SEGMENT_REPO_NAME_KEY);
+        return Objects.equals(translogRepoName, segmentRepoName);
+    }
+
+    private LatchedActionListener<Void> getUploadPathLatchedActionListener(
+        IndexMetadata indexMetadata,
+        CountDownLatch latch,
+        List<Exception> exceptionList,
+        Map<RemoteStoreEnums.DataCategory, List<RemoteStoreEnums.DataType>> pathCreationMap
+    ) {
+        return new LatchedActionListener<>(
+            ActionListener.wrap(
+                resp -> logger.trace(
+                    new ParameterizedMessage("Index path uploaded for {} indexMetadata={}", pathCreationMap, indexMetadata)
+                ),
+                ex -> {
+                    logger.error(
+                        new ParameterizedMessage(
+                            "Exception during Index path upload for {} indexMetadata={}",
+                            pathCreationMap,
+                            indexMetadata
+                        ),
+                        ex
+                    );
+                    exceptionList.add(ex);
+                }
+            ),
+            latch
+        );
+    }
+
+    /**
+     * This method checks if the index metadata has attributes that calls for uploading the index path for remote store
+     * uploads. It checks if the remote store path type is {@code HASHED_PREFIX} and returns true if so.
+     */
+    private boolean requiresPathUpload(IndexMetadata indexMetadata) {
+        // A cluster will have remote custom metadata only if the cluster is remote store enabled from data side.
+        Map<String, String> remoteCustomData = indexMetadata.getCustomData(IndexMetadata.REMOTE_STORE_CUSTOM_KEY);
```

Comment:
this custom metadata has already been pushed to main branch?

Reply:
Yes, this was added in https://github.com/opensearch-project/OpenSearch/pull/12607